content/source/docs/enterprise/sentinel/import/index.html.md

```diff
@@ -24,100 +24,178 @@ see [Mocking Terraform Sentinel Data](../mock.html).
+
+        # Loop through the resource maps, resources, and instances
+        for aws_instances as joined_path, resource_map {
+                for resource_map as name, instances {
+                        for instances as index, r {
+
+                                # Get address of the resource instance
+                                address = get_instance_address(joined_path, resource_type, name, index)
+
+                                # Validate that each instance has allowed value
+                                # If not, print violation message
+                                if r.applied.instance_type not in allowed_types {
+                                        print("EC2 instance", address, "has attribute",
+                                                r.applied.instance_type, "that is not in the list", allowed_types)
+                                        validated = false
+                                }
+                        }
+                }
+        }
+
+        return validated
+}
+```
+
+This function calls the `find_resources_from_plan` and `get_instance_address` functions described above.
+
+The boolean variable `validated` is initially set to `true`, but it is set to `false` if any resource instance violates the condition requiring that the `instance_type` attribute be in the `allowed_types` list. Since the function returns `true` or `false`, it can be called inside Sentinel rules.
+
+Note that this function prints a warning message for each resource instance that violates the condition. This allows the writer of a Sentinel rule that calls it to tell a user about **all** resource instances that violate it. This is much better than just reporting the first violation since a user with multiple violations learns about all of them in a single shot.
+
+### Call Functions From Rules
+
+Finally, having re-used the standardized `find_resources_from_plan` and `get_instance_address` functions and having written your own function to validate that resources instances of a specific type satisfy some condition, you can define a list with allowed values and write a rule that calls your function:
 
 ```python
 # Allowed Types
 allowed_types = [
-	"t2.small",
-	"t2.medium",
-	"t2.large",
+        "t2.small",
+        "t2.medium",
+        "t2.large",
 ]
 
+# Call the validation function and assign results
+instance_types_validated = validate_instance_types(allowed_types)
+
 # Rule to restrict instance types
 instance_type_allowed = rule {
-	all get_resources("aws_instance") as r {
-		r.applied.instance_type in allowed_types
-	}
+        instance_types_validated is true
+}
+
+# Main rule
+main = rule {
+        instance_type_allowed is true
+}
+
+```
+
+Note that you could leave out the `instance_type_allowed` rule and place `instance_types_validated is true` directly inside your `main` rule. You could also leave out `is true` in both rules; we included it here to be more explicit.
```

Comment:
2 issues here:

> Note that you could leave out the `instance_type_allowed` rule and place `instance_types_validated is true` directly inside your `main` rule.

If this is the case, the `instance_type_allowed` rule should just be removed from the example as it serves no extra purpose.

> You could also leave out `is true` in both rules; we included it here to be more explicit.

I get what you are saying here, but "more explicit" is a misnomer - there is no ambiguity in what a standalone unary boolean expression evaluates to.

I think that this paragraph could be removed if we correct the notes in the example:

* Remove the `instance_type_allowed` rule and just evaluate off of the `instance_types_validated` rule.
* Use the unary form of the boolean expression for `instance_types_validated`. This is readable enough and conveys the purpose just fine.

Reply:
I'll remove the rule and just use a main rule.  I actually mostly do that in most of my policies now anyways to reduce verbose Sentinel logging.